File: proton-j/proton/src/main/java/org/apache/qpid/proton/engine/impl/ConnectionImpl.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.proton.engine.impl;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import org.apache.qpid.proton.engine.*;
import org.apache.qpid.proton.type.transport.Open;

public class ConnectionImpl extends EndpointImpl implements Connection
{

    public static final int MAX_CHANNELS = 255;
    private List<SessionImpl> _sessions = new ArrayList<SessionImpl>();
    private EndpointImpl _transportTail;
    private EndpointImpl _transportHead;
    private int _maxChannels = MAX_CHANNELS;

    private LinkNode<SessionImpl> _sessionHead;
    private LinkNode<SessionImpl> _sessionTail;


    private LinkNode<LinkImpl> _linkHead;
    private LinkNode<LinkImpl> _linkTail;


    private DeliveryImpl _workHead;
    private DeliveryImpl _workTail;

    private DeliveryImpl _transportWorkHead;
    private DeliveryImpl _transportWorkTail;
    private String _localContainerId = "";
    private String _localHostname = "";
    private boolean _bound;
    private String _remoteContainer;
    private String _remoteHostname;

    public ConnectionImpl()
    {
    }

    public SessionImpl session()
    {
        SessionImpl session = new SessionImpl(this);
        _sessions.add(session);


        return session;
    }

    protected LinkNode<SessionImpl> addSessionEndpoint(SessionImpl endpoint)
    {
        LinkNode<SessionImpl> node;
        if(_sessionHead == null)
        {
            node = _sessionHead = _sessionTail = LinkNode.newList(endpoint);
        }
        else
        {
            node = _sessionTail = _sessionTail.addAtTail(endpoint);
        }
        return node;
    }

    void removeSessionEndpoint(LinkNode<SessionImpl> node)
    {
        LinkNode<SessionImpl> prev = node.getPrev();
        LinkNode<SessionImpl> next = node.getNext();

        if(_sessionHead == node)
        {
            _sessionHead = next;
        }
        if(_sessionTail == node)
        {
            _sessionTail = prev;
        }
        node.remove();
    }


    protected LinkNode<LinkImpl> addLinkEndpoint(LinkImpl endpoint)
    {
        LinkNode<LinkImpl> node;
        if(_linkHead == null)
        {
            node = _linkHead = _linkTail = LinkNode.newList(endpoint);
        }
        else
        {
            node = _linkTail = _linkTail.addAtTail(endpoint);
        }
        return node;
    }


    void removeLinkEndpoint(LinkNode<LinkImpl> node)
    {
        LinkNode<LinkImpl> prev = node.getPrev();
        LinkNode<LinkImpl> next = node.getNext();

        if(_linkHead == node)
        {
            _linkHead = next;
        }
        if(_linkTail == node)
        {
            _linkTail = prev;
        }
        node.remove();
    }


    public Session sessionHead(final EnumSet<EndpointState> local, final EnumSet<EndpointState> remote)
    {
        if(_sessionHead == null)
        {
            return null;
        }
        else
        {
            LinkNode.Query<SessionImpl> query = new EndpointImplQuery<SessionImpl>(local, remote);
            LinkNode<SessionImpl> node = query.matches(_sessionHead) ? _sessionHead : _sessionHead.next(query);
            return node == null ? null : node.getValue();
        }
    }

    public Link linkHead(EnumSet<EndpointState> local, EnumSet<EndpointState> remote)
    {
        if(_linkHead == null)
        {
            return null;
        }
        else
        {
            LinkNode.Query<LinkImpl> query = new EndpointImplQuery<LinkImpl>(local, remote);
            LinkNode<LinkImpl> node = query.matches(_linkHead) ? _linkHead : _linkHead.next(query);
            return node == null ? null : node.getValue();
        }
    }

    @Override
    protected ConnectionImpl getConnectionImpl()
    {
        return this;
    }

    public void free()
    {
        super.free();
        for(Session session : _sessions)
        {
            session.free();
        }
        _sessions = null;
    }

    public void handleOpen(Open open)
    {
        // TODO - store state
        setRemoteState(EndpointState.ACTIVE);
        setRemoteHostname(open.getHostname());
        setRemoteContainer(open.getContainerId());
    }


    EndpointImpl getTransportHead()
    {
        return _transportHead;
    }

    EndpointImpl getTransportTail()
    {
        return _transportTail;
    }

    void addModified(EndpointImpl endpoint)
    {
        if(_transportTail == null)
        {
            endpoint.setTransportNext(null);
            endpoint.setTransportPrev(null);
            _transportHead = _transportTail = endpoint;
        }
        else
        {
            _transportTail.setTransportNext(endpoint);
            endpoint.setTransportPrev(_transportTail);
            _transportTail = endpoint;
            _transportTail.setTransportNext(null);
        }
    }

    void removeModified(EndpointImpl endpoint)
    {
        if(_transportHead == endpoint)
        {
            _transportHead = endpoint.transportNext();
        }
        else
        {
            endpoint.transportPrev().setTransportNext(endpoint.transportNext());
        }

        if(_transportTail == endpoint)
        {
            _transportTail = endpoint.transportPrev();
        }
        else
        {
            endpoint.transportNext().setTransportPrev(endpoint.transportPrev());
        }
    }

    public int getMaxChannels()
    {
        return _maxChannels;
    }

    public String getLocalContainerId()
    {
        return _localContainerId;
    }

    public void setLocalContainerId(String localContainerId)
    {
        _localContainerId = localContainerId;
    }

    public DeliveryImpl getWorkHead()
    {
        return _workHead;
    }

    @Override
    public void setContainer(String container)
    {
        _localContainerId = container;
    }

    @Override
    public void setHostname(String hostname)
    {
        _localHostname = hostname;
    }

    @Override
    public String getRemoteContainer()
    {
        return _remoteContainer;
    }

    @Override
    public String getRemoteHostname()
    {
        return _remoteHostname;
    }

    public String getHostname()
    {
        return _localHostname;
    }

    void setRemoteContainer(String remoteContainerId)
    {
        _remoteContainer = remoteContainerId;
    }

    void setRemoteHostname(String remoteHostname)
    {
        _remoteHostname = remoteHostname;
    }

    DeliveryImpl getWorkTail()
    {
        return _workTail;
    }

    void removeWork(DeliveryImpl delivery)
    {
        if(_workHead == delivery)
        {
            _workHead = delivery.getWorkNext();

        }
        if(_workTail == delivery)
        {
            _workTail = delivery.getWorkPrev();
        }
    }

    void addWork(DeliveryImpl delivery)
    {
        if(_workHead != delivery && delivery.getWorkNext() == null && delivery.getWorkPrev() == null)
        {
            if(_workTail == null)
            {
                delivery.setWorkNext(null);
                delivery.setWorkPrev(null);
                _workHead = _workTail = delivery;
            }
            else
            {
                _workTail.setWorkNext(delivery);
                delivery.setWorkPrev(_workTail);
                _workTail = delivery;
                delivery.setWorkNext(null);
            }
        }
    }

    public Sequence<DeliveryImpl> getWorkSequence()
    {
        return new WorkSequence(_workHead);
    }

    public void setBound(boolean bound)
    {
        _bound = true;
    }

    public boolean isBound()
    {
        return _bound;
    }

    private class WorkSequence implements Sequence<DeliveryImpl>
    {
        private DeliveryImpl _next;

        public WorkSequence(DeliveryImpl workHead)
        {
            _next = workHead;
        }

        public DeliveryImpl next()
        {
            DeliveryImpl next = _next;
            if(next != null)
            {
                _next = next.getWorkNext();
            }
            return next;
        }
    }

    DeliveryImpl getTransportWorkHead()
    {
        return _transportWorkHead;
    }

    public void removeTransportWork(DeliveryImpl delivery)
    {
        DeliveryImpl oldHead = _transportWorkHead;
        DeliveryImpl oldTail = _transportWorkTail;
        if(_transportWorkHead == delivery)
        {
            _transportWorkHead = delivery.getTransportWorkNext();

        }
        if(_transportWorkTail == delivery)
        {
            _transportWorkTail = delivery.getTransportWorkPrev();
        }
    }


    void addTransportWork(DeliveryImpl delivery)
    {
        if(_transportWorkTail == null)
        {
            delivery.setTransportWorkNext(null);
            delivery.setTransportWorkPrev(null);
            _transportWorkHead = _transportWorkTail = delivery;
        }
        else
        {
            _transportWorkTail.setTransportWorkNext(delivery);
            delivery.setTransportWorkPrev(_transportWorkTail);
            _transportWorkTail = delivery;
            delivery.setTransportWorkNext(null);
        }
    }

    void workUpdate(DeliveryImpl delivery)
    {
        if(delivery != null)
        {
            LinkImpl link = delivery.getLink();
            if(link.workUpdate(delivery))
            {
                addWork(delivery);
            }
            else
            {
                delivery.clearWork();
            }
        }
    }
}

File: proton-c/bindings/cpp/src/connection_impl.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
#include "proton/container.hpp"
#include "proton/handler.hpp"
#include "proton/error.hpp"
#include "connection_impl.hpp"
#include "proton/transport.hpp"
#include "msg.hpp"
#include "contexts.hpp"
#include "private_impl_ref.hpp"
#include "container_impl.hpp"

#include "proton/connection.h"

namespace proton {

void connection_impl::incref(connection_impl *impl_) {
    impl_->refcount_++;
}

void connection_impl::decref(connection_impl *impl_) {
    impl_->refcount_--;
    if (impl_->refcount_ == 0)
        delete impl_;
}

connection_impl::connection_impl(class container &c, pn_connection_t &pn_conn)
    : container_(c), refcount_(0), override_(0), transport_(0), default_session_(0),
      pn_connection_(&pn_conn), reactor_reference_(this)
{
    connection_context(pn_connection_, this);
}

connection_impl::connection_impl(class container &c, handler *handler)
    : container_(c), refcount_(0), override_(0), transport_(0), default_session_(0),
      reactor_reference_(this)
{
    pn_handler_t *chandler = 0;
    if (handler) {
        container_impl *container_impl = private_impl_ref<class container>::get(c);
        chandler = container_impl->wrap_handler(handler);
    }
    pn_connection_ = pn_reactor_connection(container_.reactor(), chandler);
    if (chandler)
        pn_decref(chandler);
    connection_context(pn_connection_, this);
}

connection_impl::~connection_impl() {
    delete transport_;
    delete override_;
}

transport &connection_impl::transport() {
    if (transport_)
        return *transport_;
    throw error(MSG("connection has no transport"));
}

handler* connection_impl::override() { return override_; }
void connection_impl::override(handler *h) {
    if (override_)
        delete override_;
    override_ = h;
}

void connection_impl::open() {
    pn_connection_open(pn_connection_);
}

void connection_impl::close() {
    pn_connection_close(pn_connection_);
}

pn_connection_t *connection_impl::pn_connection() { return pn_connection_; }

std::string connection_impl::hostname() {
    return std::string(pn_connection_get_hostname(pn_connection_));
}

connection &connection_impl::connection() {
    // endpoint interface.  Should be implemented in the connection object.
    throw error(MSG("Internal error"));
}

container &connection_impl::container() { return (container_); }


// TODO: Rework this.  Rename and document excellently for user handlers on connections.
// Better: provide general solution for handlers that delete before the C reactor object
// has finished sending events.
void connection_impl::reactor_detach() {
    // "save" goes out of scope last, preventing possible recursive destructor
    // confusion with reactor_reference.
    class connection save(reactor_reference_);
    if (reactor_reference_)
        reactor_reference_ = proton::connection();
    pn_connection_ = 0;
}

connection &connection_impl::reactor_reference(pn_connection_t *conn) {
    if (!conn)
        throw error(MSG("amqp_null Proton connection"));
    connection_impl *impl_ = connection_context(conn);
    if (!impl_) {
        // First time we have seen this connection
        pn_reactor_t *reactor = pn_object_reactor(conn);
        if (!reactor)
            throw error(MSG("Invalid Proton connection specifier"));
        class container container(container_context(reactor));
        if (!container)  // can't be one created by our container
            throw error(MSG("Unknown Proton connection specifier"));
        impl_ = new connection_impl(container, *conn);
    }
    return impl_->reactor_reference_;
}

link connection_impl::link_head(endpoint::state mask) {
    return link(pn_link_head(pn_connection_, mask));
}

}
